src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -1049,6 +1057,20 @@ public InMemoryOutput downloadOutputs(RemoteAction action, RemoteActionResult re
           downloadsBuilder.add(downloadFile(action, file));
         }
       }
+    } else if (!regex.isEmpty()) {
+      for (FileMetadata file : metadata.files()) {
+        if (file.path.toString().matches(regex)) {
+          forcedDownloadsBuilder.add(downloadFile(action, file));
+        }
+      }
+
+      for (Map.Entry<Path, DirectoryMetadata> entry : metadata.directories()) {
+        for (FileMetadata file : entry.getValue().files()) {
+          if (!file.path.toString().matches(regex)) {
+            forcedDownloadsBuilder.add(downloadFile(action, file));
+          }
+        }
+      }
     } else {
       checkState(
```

Comment:
This `checkState` probably also needs to happen in the regex case? Also the `symlinks` check below. That suggests that the regex handling should be moved inside the `else` case. (I'm not sure about this.)